### PR TITLE
i#3544 RV64: Fill in the missing pieces to run actual programs

### PR DESCRIPTION
**Pull request by ksco**
This patch fills in many missing pieces related to RISC-V porting.

With this patch, plus one more patch to mangle.c yet to be submitted, we are finally able to run some programs, which means some tests can be enabled/ported for RISC-V and more RISC-V specific tests can be written.

Issue: #3544

### CODE REVIEW

**Comment by ksco** [comment]
Although our philosophy is to keep patches as small as possible, IMHO these pieces would be painful to submit individually.

**Review by abhinav92003 — APPROVED** [review]

File: core/arch/interp.c [hunk 1]
```diff
@@ -6278,8 +6279,8 @@ mangle_indirect_branch_in_trace(dcontext_t *dcontext, instrlist_t *trace,
     instr_destroy(dcontext, targeter);
     added_size -= AARCH64_INSTR_SIZE;
 
-#elif defined(ARM)
-    /* FIXME i#1551: NYI on ARM */
+#else
+    /* FIXME i#1551: NYI on ARM/RISCV64 */
     ASSERT_NOT_IMPLEMENTED(false);
 #endif /* X86/ARM */
```
[the remark this answers is not in the document]
Comment (ksco) [reply core/arch/interp.c]: Done.

File: core/arch/mangle_shared.c [hunk 2]
```diff
@@ -899,6 +899,9 @@ mangle_rseq_write_exit_reason(dcontext_t *dcontext, instrlist_t *ilist,
                            ilist, insert_at, NULL, NULL);
 #    endif
     /* FIXME i#3544: Not implemented */
```
[the remark this answers is not in the document]
Comment (ksco) [reply core/arch/mangle_shared.c]: Done.

File: core/fragment.h [hunk 3]
```diff
@@ -191,8 +191,10 @@
 #define FRAG_ISA_MODE(flags)                                                        \
     IF_X86_ELSE(                                                                    \
         IF_X64_ELSE((FRAG_IS_32(flags)) ? DR_ISA_IA32 : DR_ISA_AMD64, DR_ISA_IA32), \
-        IF_X64_ELSE(DR_ISA_ARM_A64,                                                 \
-                    (TEST(FRAG_THUMB, (flags)) ? DR_ISA_ARM_THUMB : DR_ISA_ARM_A32)))
+        IF_AARCH64_ELSE(IF_X64_ELSE(DR_ISA_ARM_A64,                                 \
```
[the remark this answers is not in the document]
Comment (ksco) [reply core/fragment.h]: Done. Sorry for the mistake.

File: core/unix/loader_linux.c [hunk 4]
```diff
@@ -260,12 +261,17 @@ privload_mod_tls_init(privmod_t *mod)
...
      * 1. increase the tls_block_size with the right alignment
      *    using ALIGN_FORWARD()
-     * 2. add first_byte to make the first byte with right alighment.
+     * 2. add first_byte to make the first byte with right alignment.
      */
     offset = first_byte +
         ALIGN_FORWARD(offset + opd->tls_block_size + first_byte, opd->tls_align);
     opd->tls_offset = offset;
     tls_info.offs[tls_info.num_mods] = offset;
+#    else
```
[the remark this answers is not in the document]
Comment (ksco) [reply core/unix/loader_linux.c]: Done.

**Comment by ksco** [comment]
x86-64-ubuntu22 api.rseq failed: https://github.com/DynamoRIO/dynamorio/issues/6185

**Comment by ksco** [comment]
x86-64 common.floatpc_xl8all  failed: https://github.com/DynamoRIO/dynamorio/issues/2267

**Comment by ksco** [comment]
x86-32 81 tests failed: https://github.com/DynamoRIO/dynamorio/issues/6417

**Comment by ksco** [comment]
x86-64 and x86-32 failed again on a second try, @abhinav92003 may I borrow your admin power? 😈

**Pull request merged**